tls/bignum.c

```diff
@@ -644,225 +633,155 @@ ttls_mpi_cmp_int(const TlsMpi *X, long z)
 	}
 
 	/* Modular comparison. */
-	return z == X->p[0] ? 0 : X->p[0] > z ? X->s : -X->s;
+	return z == MPI_P(X)[0] ? 0 : MPI_P(X)[0] > z ? X->s : -X->s;
 }
 
 /**
  * Unsigned addition: X = |A| + |B|
  *
  * @A and @B must be different, but either of them can accept the result @X.
  */
-int
+void
 ttls_mpi_add_abs(TlsMpi *X, const TlsMpi *A, const TlsMpi *B)
 {
-	size_t i, n;
-	unsigned long *a, *b, *x, c = 0;
-
 	BUG_ON(A == B);
-	if (X == B) {
+	BUG_ON(X->limbs < max_t(unsigned short, A->used, B->used));
+
+	if (B->used > A->used) {
 		const TlsMpi *T = A;
-		A = X;
+		A = B;
 		B = T;
 	}
 
+	X->used = mpi_add_x86_64(MPI_P(X), X->limbs, MPI_P(B), B->used,
+				 MPI_P(A), A->used);
+	BUG_ON(X->used <= 0);
```

Comment:
`X->used` is unsigned, so condition is never false. Either need to convert to `int` (as `mpi_add_x86_64()` return value), or use additional variable. Now the error returned by `mpi_add_x86_64()` is ignored.